src/Native/CpuMath/IntrinsicsUtils.cs

```diff
@@ -0,0 +1,796 @@
+        //        float result = 0;
+        //        for (int i = 0; i < src.Length; i++)
+        //        {
+        //            result += src[i] * dst[i];
+        //        }
+        //        return result;
+        //    }
+        //}
+
+        private static float DotUSse(float[] src, float[] dst, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+
+            unsafe
+            {
+
+                fixed (float* psrc = src)
+                fixed (float* pdst = dst)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pDstCurrent = pdst;
+                    float* pEnd = psrc + count;
+
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        result = Sse.Add(result, Sse.Multiply(srcVector, dstVector));
+
+                        pSrcCurrent += 4;
+                        pDstCurrent += 4;
+                    }
+
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
+                        result = Sse3.HorizontalAdd(result, result);
+                        result = Sse3.HorizontalAdd(result, result);
+                    }
+                    else
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
+                    }
+
+
+                    while (pSrcCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                        Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                        result = Sse.AddScalar(result, Sse.MultiplyScalar(srcVector, dstVector));
+
+                        pSrcCurrent++;
+                        pDstCurrent++;
+                    }
+                }
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+
+        public static float DotU(float[] src, float[] dst, int count)
+        {
+            if (Sse.IsSupported)
+            {
+                return DotUSse(src, dst, count);
+            }
+            else
+            {
+                // Software fallback.
+                float result = 0;
+                for (int i = 0; i < count; i++)
+                {
+                    result += src[i] * dst[i];
+                }
+                return result;
+            }
+        }
+
+        private static float DotSUSse(float[] src, float[] dst, int[] idx, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+
+            unsafe
+            {
+
```

Comment:
Another extra newline. I will stop commenting on these - just search and replace.